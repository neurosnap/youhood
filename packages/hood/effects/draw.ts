import * as L from 'leaflet';
import 'leaflet-draw';
import { call, race, put, take } from 'redux-saga/effects';

import { HoodMap } from '@youhood/map/types';

import { setEdit, hoodCreated, cancelDrawHood } from '../actions';

export function createPolygon({ map }: HoodMap) {
  return new L.Draw.Polygon(map);
}

export function* onDrawHood(hoodMap: HoodMap) {
  yield put(setEdit(true));
  const poly = yield call(createPolygon, hoodMap);
  poly.enable();

  const winner = yield race({
    cancel: take(`${cancelDrawHood}`),
    create: take(`${hoodCreated}`),
  });

  if (winner.cancel) {
    poly.disable();
    yield put(setEdit(false));
  }
}
