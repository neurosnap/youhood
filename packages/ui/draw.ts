import * as h from 'react-hyperscript';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { actions, selectors } from '@youhood/hood';
const { drawHood, cancelDrawHood } = actions;
const { getIsEditing } = selectors;
import { selectors as authSelectors } from '@youhood/auth';
const { isUserAuthenticated } = authSelectors;

import { HoodBarButton } from './ui';

const Button = HoodBarButton.extend`
  width: 91px;
  display: flex;
  justify-content: center;
`;

const Overlay = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 56px;
  background-color: rgba(0, 0, 0, 0);
  cursor: pointer;
`;

interface ICancel {
  cancelDraw: CancelFn;
}
export const Cancel: React.SFC<ICancel> = ({ cancelDraw }) =>
  h('div', [
    h(Button, { onClick: cancelDraw }, [h('div', 'Cancel')]),
    h(Overlay, { onClick: cancelDraw }),
  ]);

type CancelFn = (event: React.MouseEvent<HTMLElement>) => void;
interface IState {
  isUserLoggedIn: boolean;
  isEditing: boolean;
}
interface IDispatch {
  handleCancelDrawHood: CancelFn;
  handleDrawHood: (event: React.MouseEvent<HTMLButtonElement>) => void;
}
interface Props extends IState, IDispatch {}

export const DrawHood: React.SFC<Props> = ({
  handleDrawHood,
  handleCancelDrawHood,
  isEditing,
  isUserLoggedIn,
}) => {
  if (!isUserLoggedIn) {
    return null;
  }

  return isEditing
    ? h(Cancel, { cancelDraw: handleCancelDrawHood })
    : h(Button, { onClick: handleDrawHood }, 'Create Hood');
};

export default connect<IState, IDispatch>(
  (state) => ({
    isEditing: getIsEditing(state),
    isUserLoggedIn: isUserAuthenticated(state),
  }),
  (dispatch) => ({
    handleDrawHood: () => dispatch(drawHood()),
    handleCancelDrawHood: () => dispatch(cancelDrawHood()),
  }),
)(DrawHood);
