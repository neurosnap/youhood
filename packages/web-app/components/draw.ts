import * as h from 'react-hyperscript';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { actionCreators, selectors } from '@youhood/hood';
const { drawHood, cancelDrawHood } = actionCreators;
const { getIsEditing } = selectors;

import { HoodBarButton } from './ui';

interface Props {
  isEditing: boolean;
  handleCancelDrawHood: Function;
  handleDrawHood: Function;
}

const Button = HoodBarButton.extend`
  width: 91px;
  display: flex;
  justify-content: center;
`;

const Overlay = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 56px;
  background-color: rgba(0, 0, 0, 0);
  cursor: pointer;
`;

const Cancel = ({ cancelDraw }: { cancelDraw: Function }) =>
  h('div', [
    h(Button, { onClick: cancelDraw }, [
      h('div', 'Cancel'),
    ]),
    h(Overlay, { onClick: cancelDraw }),
  ]);

const DrawHood = ({ handleDrawHood, handleCancelDrawHood, isEditing }: Props) => isEditing ?
  h(Cancel, { cancelDraw: handleCancelDrawHood }) :
  h(Button, { onClick: handleDrawHood }, 'Create Hood');

export default connect(
  (state) => ({
    isEditing: getIsEditing(state),
  }),
  (dispatch) => ({
    handleDrawHood: () => dispatch(drawHood()),
    handleCancelDrawHood: () => dispatch(cancelDrawHood()),
  }),
)(DrawHood as any);
