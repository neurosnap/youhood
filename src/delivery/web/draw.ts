import * as h from 'react-hyperscript';
import { connect } from 'react-redux';

import { actionCreators } from '../../packages/hood';
const { drawHood } = actionCreators;

const DrawHood = ({ handleDrawHood }: { handleDrawHood: Function }) =>
  h('i.fa.fa-plus.fa-lg', { onClick: handleDrawHood });

export default connect(
  null,
  (dispatch) => ({
    handleDrawHood: () => dispatch(drawHood()),
  }),
)(DrawHood as any);
