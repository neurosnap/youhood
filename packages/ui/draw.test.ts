import { mount } from 'enzyme';
import * as h from 'react-hyperscript';

import { DrawHood } from './draw';

describe('DrawHood', () => {
  describe('when `isEditing` is true', () => {
    it('should render the cancel button', () => {
      const tree = mount(h(DrawHood, { isEditing: true }));
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when `isEditing` is false', () => {
    it('should render the create button', () => {
      const tree = mount(h(DrawHood, { isEditing: false }));
      expect(tree).toMatchSnapshot();
    });
  });
});
