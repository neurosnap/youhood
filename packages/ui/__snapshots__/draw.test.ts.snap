// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawHood when \`isEditing\` is false should render the create button 1`] = `
.c0 {
  cursor: pointer;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 10px;
  margin: 0 10px;
  background: inherit;
  color: inherit;
  border: none;
  font: inherit;
  cursor: pointer;
  outline: inherit;
  font-size: 12px;
  width: 91px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0:hover:enabled {
  background-color: rgba(255,255,255,0.25);
}

.c0:disabled {
  color: #ccc;
}

<Component
  isEditing={false}
>
  <styled.button>
    <button
      className="c0"
    >
      Create Hood
    </button>
  </styled.button>
</Component>
`;

exports[`DrawHood when \`isEditing\` is true should render the cancel button 1`] = `
.c0 {
  cursor: pointer;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 10px;
  margin: 0 10px;
  background: inherit;
  color: inherit;
  border: none;
  font: inherit;
  cursor: pointer;
  outline: inherit;
  font-size: 12px;
  width: 91px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0:hover:enabled {
  background-color: rgba(255,255,255,0.25);
}

.c0:disabled {
  color: #ccc;
}

.c1 {
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 56px;
  background-color: rgba(0,0,0,0);
  cursor: pointer;
}

<Component
  isEditing={true}
>
  <Component>
    <div>
      <styled.button>
        <button
          className="c0"
        >
          <div>
            Cancel
          </div>
        </button>
      </styled.button>
      <styled.div>
        <div
          className="c1"
        />
      </styled.div>
    </div>
  </Component>
</Component>
`;
